test/irmin-pack/test_tree.ml

```diff
@@ -144,6 +171,122 @@ let test_fold_undefined () =
   let expected = List.map fst bindings in
   test_fold ~order:`Undefined bindings expected
 
+let bin_of_proof = Irmin.Type.(unstage (to_bin_string Tree.Proof.t))
+
+let proof_of_bin s =
+  match Irmin.Type.(unstage (of_bin_string Tree.Proof.t)) s with
+  | Ok s -> s
+  | Error (`Msg e) -> Alcotest.fail e
+
+let check_completeness proof ops =
+  Lwt_list.iter_s
+    (fun k ->
+      Tree.find_tree proof k >|= function
+      | None -> Alcotest.failf "cannot read %a" Fmt.(Dump.list string) k
+      | Some _ -> ())
+    ops
+
+let check_equivalence tree proof (op, k, v) =
+  match op with
+  | `Add ->
+      let* tree = Tree.add tree k v in
+      let+ proof = Tree.add proof k v in
+      Alcotest.(check_repr Store.Hash.t)
+        (Fmt.str "same hash add %a" Fmt.(Dump.list string) k)
+        (Tree.hash tree) (Tree.hash proof);
+      (tree, proof)
+  | `Del ->
+      let* tree = Tree.remove tree k in
+      let+ proof = Tree.remove proof k in
+      Alcotest.(check_repr Store.Hash.t)
+        (Fmt.str "same hash del %a" Fmt.(Dump.list string) k)
+        (Tree.hash tree) (Tree.hash proof);
+      (tree, proof)
+
+let test_proofs steps =
+  (* init the store *)
+  let bindings = bindings steps in
```

Comment:
might be worth testing this on more complex stores (deeper + a combination of small&large directories) before merging.